File: file-handling/handling-csv-files/c-sharp/csv-library/csv_library_genre.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;
using System.Globalization;
using System.IO;
using CsvHelper;

namespace MyApplication
{
    // The Movie class holds information about a movie
    class Movie
    {
        public string Title { get; set; }
        public string Year { get; set; }
        public string Duration { get; set; }
        public string Genre { get; set; }
        public string Director { get; set; }
        public string Rating { get; set; }
        public string Revenue { get; set; }
    }

    
    class Program
    {
        // The Main method is the entry point for all C# programs
        public static void Main() {  
            Console.WriteLine("Enter a genre of movie to find:");
            string genreChosen = Console.ReadLine();

            int total = MoviesByGenre(genreChosen);
            Console.WriteLine($"There were {total} movies containing the genre {genreChosen}");
        }

        // Read all the records from a CSV file
        public static int MoviesByGenre(string genre) {
            // Initialise the filename and count
            string filename = "movies_extra.csv";
            int count = 0;

            // Capitalise the first letter of each word in genre
            TextInfo textInfo = CultureInfo.CurrentCulture.TextInfo;
            genre = textInfo.ToTitleCase(genre);

            // Read the CSV file using StreamReader and CsvReader
            using (var reader = new StreamReader(filename))
            using (var csvReader = new CsvReader(reader, CultureInfo.CurrentCulture)) {
                // Get all the records and store them as a list of Movie objects
                var records = csvReader.GetRecords<Movie>();
                
                // Loop through the list of Movie objects
                foreach (var movie in records) {
                    // Check if the genre is in the genre field
                    if (movie.Genre.Contains(genre)) {
                        Console.WriteLine($"Data: {movie.Title}, {movie.Year}, {movie.Duration}, {movie.Genre}, {movie.Director}, {movie.Rating}, {movie.Revenue}");
                        count = count + 1;
                    }
                }
            }
            return count;
        }

        
    }
}

File: file-handling/handling-csv-files/python/csv-library/csv_library_genre.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4

import csv 

def movies_by_genre(genre):
    """Display the records that match a given genre"""
    # Initialise the filename and count
    filename = "movies_extra.csv";
    count = 0
    
    # Capitalise the first letter of each word in genre
    genre = genre.title()

    # Open the file in read mode using the with command
    with open(filename, mode="r") as file_object:
        # Use the csv library to read the file and specify the delimiter
        csv_reader = csv.reader(file_object, delimiter=",")
        for line in csv_reader:
            # Check if the genre is in the genre field
            if genre in line[3]:
                print(f"Data: {line[0]}, {line[1]}, {line[2]}, {line[3]}, {line[4]}, {line[5]}, {line[6]}")
                count = count + 1
    return count


def main():
    genre_chosen = input("Enter a genre of movie to find: ")
    
    total = movies_by_genre(genre_chosen)
    print(f"There were {total} movies containing the genre {genre_chosen}")


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()
